### DISCUSSION TITLE
How do we keep Decoder lightweight?

### DISCUSSION THREAD

**Discussion by koenvervloesem** [theengs/decoder#169] [Ideas]
I saw the discussion about Decoder optimizations in https://github.com/1technophile/OpenMQTTGateway/pull/1248#issuecomment-1198820859.

Especially:

> Unfortunately, the alternative options would require a major refactor and not worth it to support outdated devices IMO.

I do think eventually we'll come to a point where this is inevitable if we want to keep Theengs Decoder lightweight, as we'll keep piling decoders on it.

Do you know [Kaitai Struct](https://kaitai.io/)? It's a code generator to create decoders for binary data in various languages. I have used it to generate Python code to decode RuuviTag BLE and iBeacon advertisements, and it's quite neat, with YAML code to describe structure of the data.

Unfortunately it hasn't been ported yet to generate C code, but [someone has been working on it](https://github.com/kaitai-io/kaitai_struct/issues/263).

Maybe this (or the code generation approach in general) is something to consider in the long run?

**Comment by 1technophile** [comment]
> I do think eventually we'll come to a point where this is inevitable if we want to keep Theengs Decoder lightweight, as we'll keep piling decoders on it.

Good to open this discussion. We can see it as a mid-term improvement, as we still have room to optimize the current architecture.

I was looking at some YAML parser but didn't dig much on how it could help:
https://github.com/jbeder/yaml-cpp

I would say that the main goal is to combine lightness and readiness regarding decoders definitions.

**Reply by DigiH** [reply]
Also, as previously discussed in https://github.com/theengs/decoder/discussions/106, introducing a **lookup table function** would greatly reduce the current [MS-CDP decoder](https://github.com/theengs/decoder/blob/development/src/devices/MS_CDP_json.h) and other future decoders benefiting from this.

**Comment by h2zero** [comment]
I guess this would depend on the definition of lightweight. 

* Reduce code space use? = reduce decoder JSON strings/objects. 
* Reduce RAM use? = break everything and use cjson or other alternatives. 
* Best of all? = custom compilation of decoder JSON files into a binary object that can be linked so only the app input JSON is required.

Starting with the first is the easiest obviously, but limited. The second could be done in a way that the applications do not need to be changed but would bloat the app, so breaking the API to use alternative JSON libraries seems necessary. The third option, as interesting as it would be, is much too extreme.

YAML may have potential, I don't know how efficient the parsers are, but would still have the API breakage problem.

**Reply by DigiH** [reply]
with part of point 1. still in your `reduce-and-optimize` branch, only giving a modest saving (~1100 bytes - at the then Decoder stage) , as stated in 

https://github.com/theengs/decoder/discussions/99

I was also wondering how exactly the `_props` definitions work, assumedly mainly for auto-discovery, and why not pool them all together in one `_common_props`

With the current Decoder 0.6.2 and OMG 0.9.15 the difference is

0.6.2
RAM:   [==        ]  17.8% (used 58456 bytes from 327680 bytes)
Flash: [=======   ]  73.9% (used 1452756 bytes from 1966080 bytes)

one single _common_props
https://github.com/DigiH/decoder/commits/properties-test
RAM:   [==        ]  17.8% (used 58448 bytes from 327680 bytes)
Flash: [=======   ]  73.5% (used 1445764 bytes from 1966080 bytes)

so another smallish saving of ~6992 bytes, but still about half of what Decoder gained since the above mentioned test.

Depending on the feasibility for auto-discovery, this could be pooled together with the above mentioned reduce-and-optimize implementations.

**Reply by h2zero** [reply]
That's a decent savings for sure but it will likely break `getTheengProperties` and force the caller to provide a huge json document buffer for it.
Would be better find make a generator for this and remove the props all together.

**Reply by DigiH** [reply]
Since each decoder already has the props defined for decoding I just tough it redundant to define the unit and name 10 times (or so) for
```
"batt":{
         "unit":"%",
         "name":"battery"
      }
```

over and over again, similar with the other repeating property definitions.

> Would be better find make a generator for this and remove the props all together.

Likely the better solution :) but I think we should take this saving opportunity - combined with your `reduce-and-optimize`; again, depending how this still allows for the correct auto-discovery of only the selected properties of each decoder device.

**Reply by 1technophile** [reply]
> Since each decoder already has the props defined for decoding I just tough it redundant to define the unit and name 10 times (or so) for

Agreed, I think we could define one key one time, and call it from the different property definitions, example in the decoder:
Lywsd03mmc_prop = temp + batt

And on a global properties dictionnary;
```
batt = "batt":{
         "unit":"%",
         "name":"battery"
      }
```
Not sure the best way to write this but you get the principle :-)

**Reply by 1technophile** [reply]
Or if we follow the thought of Digih we replace;
Lywsd03mmc_prop = temp + batt

By looking at the decoder definition. The keys listed are the ones that the getTheengsProperties return by parsing the decoder definition and taking their detailled properties into the global properties dictionary.
So that we don't break getTheengsProperties.

**Comment by h2zero** [comment]
I got thinking today and perhaps there is a compromise here. If we changed the API to accept a `char*` for input and return value, which would point to a JSON string, we could use any JSON lib (or roll our own slim version). That combined with the ideas above could possibly reduce the lib size and RAM by a significant amount.

**Reply by h2zero** [reply]
For interest sake I thought I'd see what the difference was between cJSON and ArduinoJSON.
Here are the results:
```
Model: MiFlora, cJSON size: 634, ArduinoJSON size: 1783
Model: Cleargrass clock, cJSON size: 524, ArduinoJSON size: 1499
Model: Mi Jia round, cJSON size: 640, ArduinoJSON size: 1822
Model: CG_THP, cJSON size: 391, ArduinoJSON size: 1124
Model: CG_round_v1, cJSON size: 408, ArduinoJSON size: 1183
Model: CG_round_v2, cJSON size: 495, ArduinoJSON size: 1398
Model: Air Monitor Lite, cJSON size: 565, ArduinoJSON size: 1661
Model: CG_alarm_clock, cJSON size: 379, ArduinoJSON size: 1120
Model: TH lite, cJSON size: 334, ArduinoJSON size: 985
Model: Door sensor, cJSON size: 494, ArduinoJSON size: 1488
Model: Formaldehyde detector, cJSON size: 625, ArduinoJSON size: 1733
Model: TH Sensor, cJSON size: 383, ArduinoJSON size: 1067
Model: TH Sensor, cJSON size: 470, ArduinoJSON size: 1326
Model: iBBQ, cJSON size: 742, ArduinoJSON size: 2096
Model: iBBQ/SOLIS6, cJSON size: 1057, ArduinoJSON size: 2906
Model: Mi Band/Smart Watch, cJSON size: 327, ArduinoJSON size: 872
Model: Mi_Smart_Scale, cJSON size: 671, ArduinoJSON size: 2021
Model: Mi_Body_Composition_Scale, cJSON size: 744, ArduinoJSON size: 2186
Model: TPMS, cJSON size: 576, ArduinoJSON size: 1557
Model: LYWSD03MMC, cJSON size: 445, ArduinoJSON size: 1192
Model: Motion & Light, cJSON size: 379, ArduinoJSON size: 1035
Model: WS02, cJSON size: 412, ArduinoJSON size: 1131
Model: WS08, cJSON size: 412, ArduinoJSON size: 1131
Model: Bluetooth BBQ Thermometer, cJSON size: 1487, ArduinoJSON size: 4471
Model: Smart Thermo Hygrometer, cJSON size: 449, ArduinoJSON size: 1292
Model: Thermo Hygrometer, cJSON size: 443, ArduinoJSON size: 1286
Model: Smart Thermo Hygrometer, cJSON size: 449, ArduinoJSON size: 1292
Model: LYWSD03MMC, cJSON size: 473, ArduinoJSON size: 1315
Model: MiLamp, cJSON size: 243, ArduinoJSON size: 658
Model: Beacon, cJSON size: 392, ArduinoJSON size: 1040
Model: BeaconX Pro, cJSON size: 1064, ArduinoJSON size: 2909
Model: Energy Meter, cJSON size: 1140, ArduinoJSON size: 3351
Model: iBBQ, cJSON size: 528, ArduinoJSON size: 1617
Model: RuuviTag, cJSON size: 832, ArduinoJSON size: 2292
Model: RuuviTag, cJSON size: 1136, ArduinoJSON size: 3200
Model: Contact Sensor, cJSON size: 716, ArduinoJSON size: 2128
Model: Motion Sensor, cJSON size: 1115, ArduinoJSON size: 3508
Model: Meter (Plus), cJSON size: 808, ArduinoJSON size: 2577
Model: Bot, cJSON size: 386, ArduinoJSON size: 1220
Model: Laundry Sensor, cJSON size: 650, ArduinoJSON size: 1787
Model: TempoDisc, cJSON size: 493, ArduinoJSON size: 1294
Model: MS-CDP, cJSON size: 1132, ArduinoJSON size: 2613
Model: GAEN, cJSON size: 229, ArduinoJSON size: 624
Model: RoPot, cJSON size: 323, ArduinoJSON size: 875
Model: b-parasite, cJSON size: 861, ArduinoJSON size: 2584
Model: BM2 Battery Monitor, cJSON size: 227, ArduinoJSON size: 611
Model: RDL52832, cJSON size: 2784, ArduinoJSON size: 8124
Model: iBeacon, cJSON size: 675, ArduinoJSON size: 1780
```

So cJSON consumes approximately 30% of the RAM that ArduinoJSON uses, that's not chump change 😄 
For some reason there is an issue with the switchbot curtain json though that I didn't bother looking into so it was commented out.

**Reply by 1technophile** [reply]
Interesting that's a huge win in size, I found also this https://arduinojson.org/v6/how-to/reduce-memory-usage/ as our issue is mainly with esp8266 should we store long string in flash?

**Reply by DigiH** [reply]
Will have a look to see what might be the issue with the SwitchBot Curtain.

**Reply by DigiH** [reply]
Found a missed masking optimisation, but other than that I can't see anything obviously wrong with the JSON there.

https://github.com/theengs/decoder/pull/176/files

This roughly 70% win with cJSON really is something we should take, giving us the chance to get at least 3 times as many devices as currently into Decoder until we might end up in the current situation again :wink:

**Reply by h2zero** [reply]
> Will have a look to see what might be the issue with the SwitchBot Curtain.

I didn't see any issues with it either, just that cJSON wouldn't parse it, not sure why, wasn't really important for this quick test.

It's not so much about the number of devices, but the size of the biggest JSON, which is `Model: RDL52832, cJSON size: 2784, ArduinoJSON size: 8124`. This determines how much RAM we need to allocate for all the decoder JSON's. Adding more devices just consumes more flash/CPU cycles, which is the least of our issues at the moment.

Whether this is worth pursuing or not from here I can't say (big job) but it's nice to have the data.

> I found also this https://arduinojson.org/v6/how-to/reduce-memory-usage/ as our issue is mainly with esp8266 should we store long string in flash?

I'll test that out and see what difference it makes.

**Reply by DigiH** [reply]
> It's not so much about the number of devices, but the size of the biggest JSON

True ;) but then allowing for some decoders to be extended, whereas in the past we deliberately left out some - not so important - properties, which could now be included for completeness sake.

**Reply by h2zero** [reply]
Some more interesting discovery; after @1technophile mentioned this https://arduinojson.org/v6/how-to/reduce-memory-usage I decided to see what happens when using PROGMEM as a decorator for the json's.

With `[env:nodemcuv2-ble]` The results are very, very encouraging:
```
RAM:   [======    ]  56.0% (used 45852 bytes from 81920 bytes)
Flash: [======    ]  61.3% (used 640244 bytes from 1044464 bytes)
```
